### PR TITLE
fixes #21280; Enum with int64.high() value crashes compiler

### LINKED ISSUE #21280: Enum with int64.high() value crashes compiler
### Description

```nim
type
  Test = enum
    A = int64.high()
```

### Nim Version

Nim Compiler Version 1.7.1 [Linux: amd64]
Compiled at 2022-07-17
Copyright (c) 2006-2022 by Andreas Rumpf

### Current Output

```text
fatal.nim(53)            sysFatal
Error: unhandled exception: over- or underflow [OverflowDefect]
```

### Expected Output

```text
Compiles
```

### Possible Solution

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by ringabout**
fixes #21280

### CODE REVIEW

File: compiler/semtypes.nim [hunk 1]
```diff
@@ -147,7 +148,11 @@ proc semEnum(c: PContext, n: PNode, prev: PType): PType =
         declarePureEnumField(c, e)
     if (let conflict = strTableInclReportConflict(symbols, e); conflict != nil):
       wrongRedefinition(c, e.info, e.name.s, conflict.info)
-    inc(counter)
+    if counter == high(typeof(counter)):
+      if i > 1 and result.n[i-2].sym.position == high(int):
```
Comment (Araq) [inline compiler/semtypes.nim:152]: Why is the logic not as simple as

```suggestion
    if counter == high(typeof(counter)):
```

Comment (ringabout) [reply compiler/semtypes.nim:152]: It is used to take care of this case
```nim
type
  Test = enum
    A = int64.high()
    B
```
If the counter stays the same as the maximum, it won't report overflow errors for this case and A and B will have the same value.

File: tests/enum/tenum.nim [hunk 2]
```diff
@@ -175,3 +175,11 @@ block: # bug #12589
...
 
     doAssert $typ() == "wkbPoint25D"
+
+block: # bug #21280
+  type
+    Test = enum
+      B = 19
+      A = int64.high()
+
+  doAssert ord(A) == int64.high()
```
Comment (ringabout) [inline tests/enum/tenum.nim]:
```suggestion
  when not defined(i386):
    doAssert ord(A) == int64.high()
  else:
    discard A
```

File: tests/enum/tenum.nim [hunk 3]
```diff
@@ -175,3 +175,14 @@ block: # bug #12589
...
 
     doAssert $typ() == "wkbPoint25D"
+
+block: # bug #21280
+  type
+    Test = enum
+      B = 19
+      A = int64.high()
+
+  when not defined(i386):
```
Comment (ringabout) [inline tests/enum/tenum.nim]: Since `ord` returns int, it probably should returns int64.

File: tests/enum/tenum.nim [hunk 4]
```diff
@@ -175,3 +175,14 @@ block: # bug #12589
...
+    doAssert ord(A) == int64.high()
+  else:
+    discard A
```
Comment (ringabout) [inline tests/enum/tenum.nim]:
```suggestion
when not defined(i386):
  block: # bug #21280
    type
      Test = enum
        B = 19
        A = int64.high()
  
    doAssert ord(A) == int64.high()
```

File: tests/enum/tenum.nim [hunk 5]
```diff
@@ -174,5 +174,15 @@ block: # bug #12589
     proc typ(): OGRwkbGeometryType =
       return wkbPoint25D
 
+    doAssert $typ() == "wkbPoint25D"
+
```
Comment (ringabout) [inline tests/enum/tenum.nim]:
```suggestion
```

**Pull request merged**